package/network/utils/umbim/files/lib/netifd/proto/mbim.sh

```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
+	tid=$((tid + 1))
+
 	uci_set_state network $interface tid "$tid"
 
-	echo "mbim[$$]" "Connected, starting DHCP"
+	local ip_4 ip_6
+	ip_4=$(awk '$1=="ipv4address:" {print $2}' "$config")
+	ip_6=$(awk '$1=="ipv6address:" {print $2}' "$config")
+	[ "$ip_4" ] || [ "$ip_6" ] || {
+		echo "mbim[$$]" "Failed to obtain IP addresses"
+		proto_notify_error "$interface" CONFIG_FAILED
+		rm -f "$config"
+		return 1
+	}
+
 	proto_init_update "$ifname" 1
-	proto_send_update "$interface"
+	proto_set_keep 1
+	local ip mask gateway mtu dns1 dns2
+
+	[ "$ip_4" ] && {
+		echo "mbim[$$]" "Configure IPv4 on $ifname"
+		ip=${ip_4%%/*}
+		mask=${ip_4##*/}
+		gateway=$(awk '$1=="ipv4gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv4mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "1p")
+		dns2=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv4_address "$ip" "$mask"
+		[ "$defaultroute" = 0  ] || proto_add_ipv4_route 0.0.0.0 0 "$gateway" "$ip_4" "$metric"
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+	}
+	[ "$ip_6" ] && {
+		echo "mbim[$$]" "Configure IPv6 on $ifname"
+		ip=${ip_6%%/*}
+		mask=${ip_6##*/}
+		gateway=$(awk '$1=="ipv6gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv6mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "1p")
+		dns2=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv6_address "$ip" "$mask"
+		proto_add_ipv6_prefix "$ip_6"
+		[ "$defaultroute" = 0  ] || proto_add_ipv6_route "::" 0 "$gateway" "" "" "$ip_6" "$metric"
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+		# TODO: Don't know how to propagate IPv6 address to lan from here?
+	}
 
-	json_init
-	json_add_string name "${interface}_4"
-	json_add_string ifname "@$interface"
-	json_add_string proto "dhcp"
-	proto_add_dynamic_defaults
-	json_close_object
-	ubus call network add_dynamic "$(json_dump)"
-
-	json_init
-	json_add_string name "${interface}_6"
-	json_add_string ifname "@$interface"
-	json_add_string proto "dhcpv6"
-	json_add_string extendprefix 1
-	proto_add_dynamic_defaults
-	ubus call network add_dynamic "$(json_dump)"
+	rm -f "$config"
+	proto_send_update "$interface"
+	echo "mbim[$$]" "Connection setup complete"
 }
 
 proto_mbim_setup() {
 	local ret
 
-	_proto_mbim_setup $@
+	_proto_mbim_setup "$@"
 	ret=$?
 
 	[ "$ret" = 0 ] || {
 		logger "mbim bringup failed, retry in 15s"
 		sleep 15
 	}
 
-	return $rt
+	return $ret
 }
 
 proto_mbim_teardown() {
 	local interface="$1"
 
-	local device
+	local device tid
 	json_get_vars device
-	local tid=$(uci_get_state network $interface tid)
+	tid=$(uci_get_state network "$interface" tid)
```

Comment:
This is actually unrelated to your PR but the use of uci_get_stage or uci_set_stage should be considered deprecated as the underlying uci state mechanism is deprecated as well. It would be better to store the `tid` in the ubus interface data section. Example of using interface data can be seen here: https://github.com/openwrt/openwrt/blob/master/package/network/config/netifd/files/lib/netifd/dhcp.script#L49

Reply:
done, i think